packages/analytics/amplify_analytics_pinpoint_dart/lib/src/impl/analytics_client/event_client/event_client.dart

```diff
@@ -0,0 +1,214 @@
+    );
+
+    final batchItems = BuiltMap<String, EventsBatch>(
+      {_fixedEndpointId: batch},
+    );
+
+    try {
+      final result = await _pinpointClient.putEvents(
+        PutEventsRequest(
+          applicationId: _appId,
+          eventsRequest: EventsRequest(batchItem: batchItems),
+        ),
+      );
+
+      // Parse the EndpointResponse portion of Result
+      final endpointResponse = result.eventsResponse.results?[_fixedEndpointId];
+
+      if (endpointResponse == null) {
+        _logger.warn(
+          'putEvents - no EndpointResponse object received from Pinpoint',
+        );
+      }
+
+      // Parse the EndpointItemResponse portion of Result
+      final endpointItemResponse = endpointResponse?.endpointItemResponse;
+
+      if (endpointItemResponse == null) {
+        _logger.warn(
+          'putEvents - no PinpointEndpoint response received from Pinpoint',
+        );
+      }
+      if (endpointItemResponse?.statusCode != 202) {
+        _logger.warn(
+          'putEvents - issue with PinpointEndpoint response: $endpointItemResponse',
+        );
+      }
+
+      // Parse the individual response for each Event in batch
+      final eventsItemResponse = endpointResponse?.eventsItemResponse;
+
+      if (eventsItemResponse == null) {
+        _logger.warn(
+          'putEvents - no EventsItemResponse response received from Pinpoint',
+        );
+      }
+      eventsItemResponse?.forEach((eventID, eventItemResponse) {
+        if (!_equalsIgnoreCase(eventItemResponse.message ?? '', 'Accepted')) {
+          _logger.warn(
+            'putEvents - issue with eventId: $eventID \n $eventItemResponse',
+          );
+
+          if (_isRetryable(eventItemResponse.message ?? '')) {
+            _logger.warn(
+              'putEvents - recoverable issue, will attempt to resend: $eventID in next FlushEvents',
+            );
+            eventIdsToDelete.remove(eventID);
+          }
+        }
+      });
+    } on BadRequestException catch (e) {
+      _handleRecoverableException(e, eventIdsToDelete);
+    } on InternalServerErrorException catch (e) {
+      _handleRecoverableException(e, eventIdsToDelete);
+    } on NotFoundException catch (e) {
+      _handleRecoverableException(e, eventIdsToDelete);
+    } on TooManyRequestsException catch (e) {
+      _handleRecoverableException(e, eventIdsToDelete);
+    } on PayloadTooLargeException catch (e) {
+      _logger
+        ..warn('putEvents - exception encountered: ', e)
+        ..warn(
+            'Pinpoint event batch limits exceeded: 100 events / 4 mb total size / 1 mb max size per event \n '
+            'Reduce your event size or change number of events in a batch')
+        ..warn('Unrecoverable issue, deleting cache for local event batch');
+    } on Exception catch (e) {
+      _logger
+        ..warn('putEvents - exception encountered: ', e)
+        ..warn('Unrecoverable issue, deleting cache for local event batch');
```

Comment:
Why are we deleting events for some exceptions? I feel like this should be the opposite.. we should not be deleting IDs by default and only deleting events for very particular cases.

Reply:
Some exceptions are not recoverable.  If the exception is recoverable we do not delete those events (by clearing eventIdsToDelete) in _handleRecoversableException. 

The logic is that future requests will continue to fail due to what is contained within the event batch.  For example PayloadTooLargeException or any other Exception type that is not recoverable.

Reply:
The following exception types are defined in the Pinpoint SDK: 

Treated as recoverable: 
BadRequestException 
InternalServerErrorException
NotFoundException
TooManyRequestsException

Treated as non recoverable (events are deleted): 
ForbiddenException 
MethodNotAllowedException
PayloadTooLargeException

Reply:
Okay, so my point is we should be catching those errors and only deleting on those exceptions.

Why is `ForbiddenException` unrecoverable? Couldn't it potentially be solved by signing in (e.g. if unauthenticated identities are not allowed)?

I don't see why deleting events would help with a `MethodNotAllowedException`.